core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -679,6 +679,22 @@ private object RTS {
                         case Async.MaybeLaterIO(c) => Async.MaybeLaterIO(c)
                       }
                     }
+
+                  case IO.Tags.Ensuring =>
+                    val io = curIo.asInstanceOf[IO.Ensuring[E, Any]]
+
+                    val ref = new AtomicReference[Boolean](false)
+
+                    val finalizer = Finalizer[E](
+                      _ => IO.suspend[Void, Unit](if (!ref.get) io.finalizer else IO.unit)
+                    )
+
+                    stack.push(finalizer)
+
+                    curIo = for {
+                      a <- io.io
+                      _ <- io.finalizer[E] *> IO.sync[E, Unit](ref.set(true))
+                    } yield a
```

Comment:
I forgot this crazy complication would be necessary before Phase 3.

In the new version, it will be simpler. 

```scala
val io = curIo.asInstanceOf[IO.Ensuring[E, Any]]

val finalizer = Finalizer[E](io.finalizer)

stack.push(finalizer)

curIo = io.io
```

Also:

```scala
class Finalizer[E](finalizer: IO[Void, Unit]) extends Any => IO[E, Any] {
  def apply(a: Any): IO[E, Any] = finalizer *> IO.now(a)
}
```

Reply:
That's much more elegant indeed! While writing the code I knew what I had to do, but I still don't get the design as a whole. More reading to do. Has the paper "Concurrent Haskell" been a source of inspiration to you? I've had my eyes on it for a while, it's time I allocate some time for it.

I _think_ I need to make the second action uninterruptible, just in case something happens while in `*>`.